fdbclient/AuditUtils.actor.cpp

```diff
@@ -45,6 +49,87 @@ ACTOR static Future<std::vector<AuditStorageState>> getLatestAuditStatesImpl(Tra
 	return auditStates;
 }
 
+ACTOR Future<Void> clearAuditMetadata(Database cx, AuditType auditType, UID auditId, bool clearProgressMetadata) {
+	state Transaction tr(cx);
+	TraceEvent(SevDebug, "ClearAuditMetadataStart", auditId).detail("AuditKey", auditKey(auditType, auditId));
+
+	loop {
+		try {
+			tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+			tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			// Clear audit
+			tr.clear(auditKey(auditType, auditId));
+			if (clearProgressMetadata) {
+				// Here, we do not distinguish which range exactly used by an audit
+				// Simply clear any possible place of storing the progress of the audit
+				tr.clear(auditServerBasedProgressRangeFor(auditType, auditId));
+				tr.clear(auditRangeBasedProgressRangeFor(auditType, auditId));
+			}
+			wait(tr.commit());
+			break;
+		} catch (Error& e) {
+			TraceEvent(SevInfo, "ClearAuditMetadataError", auditId)
+			    .errorUnsuppressed(e)
+			    .detail("AuditKey", auditKey(auditType, auditId));
+			wait(tr.onError(e));
+		}
+	}
+	TraceEvent(SevDebug, "ClearAuditMetadataEnd", auditId).detail("AuditKey", auditKey(auditType, auditId));
+
+	return Void();
+}
+
+ACTOR Future<Void> clearAuditMetadataForType(Database cx, AuditType auditType, int numFinishAuditToKeep) {
+	state Transaction tr(cx);
+	state int numFinishAuditCleaned = 0; // We regard "Complete" and "Failed" audits as finish audits
+	TraceEvent(SevDebug, "DDClearAuditMetadataForTypeStart").detail("AuditType", auditType);
+
+	loop {
```

Comment:
In order to get the keys from beginning to the last N, I feel you could use `getRange` with `KeySelector`: 
```
getRange(const KeySelector& begin,
	                 const KeySelector& end,
	                 GetRangeLimits limits,
	                 Snapshot = Snapshot::False,
	                 Reverse = Reverse::False);

Reply:
Currently, getAuditStatesImpl is implemented by:
`Future<RangeResult> getRange(const KeyRange& keys,
	                                           int limit,
	                                           Snapshot snapshot = Snapshot::False,
	                                           Reverse reverse = Reverse::False) {
		return getRange(KeySelector(firstGreaterOrEqual(keys.begin), keys.arena()),
		                KeySelector(firstGreaterOrEqual(keys.end), keys.arena()),
		                limit,
		                snapshot,
		                reverse);
	}`